---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

```{r, include=FALSE}
library(magrittr)
```

# parcutils

<!-- badges: start -->

```{r , echo=FALSE , results='asis' , message=FALSE}
cat(
badger::badge_devel("cparsania/parcutils" , color = "blue"),
badger::badge_lifecycle()
)
```

<!-- badges: end -->

The goal of parcutils is to provide day to day bioinformatics utility functions. Those functions are mostly wrapper functions written on top of core bioconductior fucntions. 

## Installation


```{r, eval=FALSE}

if(require("devtools")){
        devtools::install_github("cparsania/parcutils")
} else{
        install.packages("devtools")
        devtools::install_github("cparsania/pparcutils")
}
```

## Example

This is a basic example which shows you how to solve a common problem:

## Intergenic regions and signal intensity

Given the .gff file and .bw file, identify genome wide intergenic regions and their signal intensity. 

```{r}
library(parcutils)


bw_file <- system.file("extdata" , "example.bw" , package = "parcutils")
gff_file <- system.file("extdata" , "C_glabrata_CBS138_version_s02-m07-r06_features.gff" ,package = "parcutils")

oo <- parcutils::get_intergenic_singals(bw_file = bw_file,gff_file  = gff_file)

oo 
```

Export results

```{r,eval=FALSE }
## export bed file 

oo %>% rtracklayer::export.bed(con = file(paste("intergenic", ".bed",sep = "")))

## export bdg file

oo %>% rtracklayer::export.bedGraph(con = file(paste("intergenic", ".bdg",sep = "")))

```

Below is the IGV snapshot showing genes(blue), intergenic regions (red) and intergenic intensities (green).

![](man/figures/intergenic_snapshot.png)







