<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="The package contains utility functions to deal with day to
    day bioinformatics tasks. Most of the functions are related to RNA-seq data analysis and visualization. ">
<title>Utility Functions To Deal Day To Day Bioinformatics Tasks. • parcutils</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Utility Functions To Deal Day To Day Bioinformatics Tasks.">
<meta property="og:description" content="The package contains utility functions to deal with day to
    day bioinformatics tasks. Most of the functions are related to RNA-seq data analysis and visualization. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">parcutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cparsania/parcutils/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="parcutils">parcutils<a class="anchor" aria-label="anchor" href="#parcutils"></a>
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<p>The goal of <code>parcutils</code> is to provide day to day bioinformatics utility functions. Most of the functions in the package are useful for analyzing and visualizing complex RNA-seq studies.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="kw">if</span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span> <span class="op">&amp;&amp;</span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span> <span class="op">)</span>
  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cparsania/parcutils"</span><span class="op">)</span>
<span class="op">}</span> <span class="kw">else</span><span class="op">{</span>
  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"devtools"</span>,<span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span>
  <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span> <span class="op">)</span>
  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cparsania/parcutils"</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="rna-seq-analysis">RNA-seq analysis<a class="anchor" aria-label="anchor" href="#rna-seq-analysis"></a>
</h2>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<div class="section level4">
<h4 id="prepare-a-count-table">Prepare a count table<a class="anchor" aria-label="anchor" href="#prepare-a-count-table"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">count_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"toy_counts.txt"</span> , package <span class="op">=</span> <span class="st">"parcutils"</span><span class="op">)</span>

<span class="va">count_data</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_delim</a></span><span class="op">(</span><span class="va">count_file</span>, delim <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span>

<span class="va">count_data</span> 
<span class="co">#&gt; # A tibble: 5,000 × 10</span>
<span class="co">#&gt;    gene_id        control_rep1 control_rep2 control_rep3 treat1_rep1 treat1_rep2</span>
<span class="co">#&gt;    &lt;chr&gt;                 &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;</span>
<span class="co">#&gt;  1 ENSG000001735…            0            0            0           0           0</span>
<span class="co">#&gt;  2 ENSG000001063…            1            0            2           0           0</span>
<span class="co">#&gt;  3 ENSG000001311…            2            0            2           0           0</span>
<span class="co">#&gt;  4 ENSG000001543…          652          690          639         607         453</span>
<span class="co">#&gt;  5 ENSG000001964…         3372         3631         3188        4644        3168</span>
<span class="co">#&gt;  6 ENSG000001730…          694          784          829         974         580</span>
<span class="co">#&gt;  7 ENSG000001405…           87           73           81         100          77</span>
<span class="co">#&gt;  8 ENSG000001875…            4            4            2           0           5</span>
<span class="co">#&gt;  9 ENSG000001392…         1374         1789         1564        1933        1459</span>
<span class="co">#&gt; 10 ENSG000001658…         3639         4533         3921        3879        3500</span>
<span class="co">#&gt; # … with 4,990 more rows, and 4 more variables: treat1_rep3 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   treat2_rep1 &lt;dbl&gt;, treat2_rep2 &lt;dbl&gt;, treat2_rep3 &lt;dbl&gt;</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="group-replicates-by-samples">Group replicates by samples<a class="anchor" aria-label="anchor" href="#group-replicates-by-samples"></a>
</h4>
<p>To run DESeq2 sample replicates replicates needs to be grouped by sample.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sample_info</span> <span class="op">&lt;-</span> <span class="va">count_data</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">.</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>  <span class="op">%&gt;%</span>
 <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="va">.</span> , groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span> , <span class="st">"treatment2"</span><span class="op">)</span> , 
                                           each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>
<span class="va">sample_info</span>
<span class="co">#&gt; # A tibble: 9 × 2</span>
<span class="co">#&gt;   samples      groups    </span>
<span class="co">#&gt;   &lt;chr&gt;        &lt;chr&gt;     </span>
<span class="co">#&gt; 1 control_rep1 control   </span>
<span class="co">#&gt; 2 control_rep2 control   </span>
<span class="co">#&gt; 3 control_rep3 control   </span>
<span class="co">#&gt; 4 treat1_rep1  treatment1</span>
<span class="co">#&gt; 5 treat1_rep2  treatment1</span>
<span class="co">#&gt; 6 treat1_rep3  treatment1</span>
<span class="co">#&gt; 7 treat2_rep1  treatment2</span>
<span class="co">#&gt; 8 treat2_rep2  treatment2</span>
<span class="co">#&gt; 9 treat2_rep3  treatment2</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="run-deseq2-for-multiple-differential-gene-comparison">Run <code>DESeq2</code> for multiple differential gene comparison.<a class="anchor" aria-label="anchor" href="#run-deseq2-for-multiple-differential-gene-comparison"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/run_deseq_analysis.html">run_deseq_analysis</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_data</span> ,
                         sample_info <span class="op">=</span> <span class="va">sample_info</span>,
                         column_geneid <span class="op">=</span> <span class="st">"gene_id"</span> ,
                         log2fc_cutoff <span class="op">=</span> <span class="fl">0.6</span>,
                         cutoff_pval <span class="op">=</span> <span class="fl">0.05</span>,
                         group_numerator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1"</span>, <span class="st">"treatment2"</span><span class="op">)</span> ,
                         group_denominator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span><span class="op">)</span>,
                         column_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control_rep1"</span>, <span class="st">"treat1_rep1"</span>, <span class="st">"treat2_rep1"</span>, <span class="st">"control_rep2"</span>, <span class="st">"treat1_rep2"</span>, <span class="st">"treat2_rep2"</span>, <span class="st">"control_rep3"</span>, <span class="st">"treat1_rep3"</span>, <span class="st">"treat2_rep3"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level4">
<h4 id="lets-have-a-look-in-to-res">Let’s have a look in to <code>res</code>
<a class="anchor" aria-label="anchor" href="#lets-have-a-look-in-to-res"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Total number of genes used for DEG analysis are 4034. </span>
<span class="co">#&gt; Total number of comparisons are 2. </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Number of DEGs in each comparison:</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; treatment1_VS_control </span>
<span class="co">#&gt; • number of up genes   : 13.</span>
<span class="co">#&gt; • number of down genes : 28.</span>
<span class="co">#&gt; ──────────────────────────────</span>
<span class="co">#&gt; treatment2_VS_control </span>
<span class="co">#&gt; • number of up genes   : 333.</span>
<span class="co">#&gt; • number of down genes : 502.</span>
<span class="co">#&gt; ──────────────────────────────</span></code></pre></div>
<p><code>res</code> is an object of improved <code>dataframe</code> - <code>tibble</code>. Each row in the <code>res</code> is a differential comparison which can be identified by the value from the column <code>comp</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span><span class="op">$</span><span class="va">comp</span>
<span class="co">#&gt; [1] "treatment1_VS_control" "treatment2_VS_control"</span></code></pre></div>
<p>Data related to each differential comparison can be found from other columns of <code>res</code>.</p>
<p>For example, summary of differently expressed genes can be found from the column <code>deg_summmary</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span><span class="op">$</span><span class="va">deg_summmary</span>
<span class="co">#&gt; $treatment1_VS_control</span>
<span class="co">#&gt; # A tibble: 3 × 2</span>
<span class="co">#&gt;   regul     n</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1 Down     28</span>
<span class="co">#&gt; 2 other  3993</span>
<span class="co">#&gt; 3 Up       13</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $treatment2_VS_control</span>
<span class="co">#&gt; # A tibble: 3 × 2</span>
<span class="co">#&gt;   regul     n</span>
<span class="co">#&gt;   &lt;chr&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1 Down    502</span>
<span class="co">#&gt; 2 other  3199</span>
<span class="co">#&gt; 3 Up      333</span></code></pre></div>
<p>As described below there are several helper functions to get data from the <code>res</code> .</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="get-data-from-res-using-helper-functions">Get data from <code>res</code> using helper functions<a class="anchor" aria-label="anchor" href="#get-data-from-res-using-helper-functions"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get normalised gene expression value for all genes across all samples. </span>
<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_normalised_expression_matrix.html">get_normalised_expression_matrix</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                            samples <span class="op">=</span> <span class="cn">NULL</span>,
                                            genes <span class="op">=</span> <span class="cn">NULL</span>,
                                            summarise_replicates <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 4,034 × 10</span>
<span class="co">#&gt;    gene_id         treat1_rep1 treat1_rep2 treat1_rep3 control_rep1 control_rep2</span>
<span class="co">#&gt;    &lt;chr&gt;                 &lt;dbl&gt;       &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;        &lt;dbl&gt;</span>
<span class="co">#&gt;  1 ENSG0000015434…      440.        461.       456.           694.        687.  </span>
<span class="co">#&gt;  2 ENSG0000019641…     3370.       3222.      3478.          3591.       3616.  </span>
<span class="co">#&gt;  3 ENSG0000017302…      707.        590.       709.           739.        781.  </span>
<span class="co">#&gt;  4 ENSG0000014059…       72.6        78.3       71.3           92.7        72.7 </span>
<span class="co">#&gt;  5 ENSG0000013926…     1403.       1484.      1568.          1463.       1781.  </span>
<span class="co">#&gt;  6 ENSG0000016589…     2815.       3559.      3404.          3875.       4514.  </span>
<span class="co">#&gt;  7 ENSG0000017945…    12679.      12542.     13690.         13257.      13358.  </span>
<span class="co">#&gt;  8 ENSG0000018886…      461.        434.       459.           540.        592.  </span>
<span class="co">#&gt;  9 ENSG0000024915…        0           2.03       0.990          0          11.0 </span>
<span class="co">#&gt; 10 ENSG0000011085…        2.18        9.15       9.90          11.7         5.97</span>
<span class="co">#&gt; # … with 4,024 more rows, and 4 more variables: control_rep3 &lt;dbl&gt;,</span>
<span class="co">#&gt; #   treat2_rep1 &lt;dbl&gt;, treat2_rep2 &lt;dbl&gt;, treat2_rep3 &lt;dbl&gt;</span>

<span class="co"># average gene expression values across relicates  </span>
<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_normalised_expression_matrix.html">get_normalised_expression_matrix</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                            samples <span class="op">=</span> <span class="cn">NULL</span>,
                                            genes <span class="op">=</span> <span class="cn">NULL</span>,
                                            summarise_replicates <span class="op">=</span> <span class="cn">T</span>, 
                                            summarise_method <span class="op">=</span> <span class="st">"median"</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 4,034 × 4</span>
<span class="co">#&gt;    gene_id                  control treatment1 treatment2</span>
<span class="co">#&gt;    &lt;chr&gt;                      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;</span>
<span class="co">#&gt;  1 ENSG00000000971:CFH       7349.      5727.      9849. </span>
<span class="co">#&gt;  2 ENSG00000001461:NIPAL3     134.       126.        68.2</span>
<span class="co">#&gt;  3 ENSG00000001497:LAS1L     9282.      8228.      7678. </span>
<span class="co">#&gt;  4 ENSG00000001631:KRIT1    23294.     29183.     20375. </span>
<span class="co">#&gt;  5 ENSG00000002746:HECW1       66.0       31.9       24.9</span>
<span class="co">#&gt;  6 ENSG00000003056:M6PR      6538.      7394.      6694. </span>
<span class="co">#&gt;  7 ENSG00000003436:TFPI       248.       225.       200. </span>
<span class="co">#&gt;  8 ENSG00000003509:NDUFAF7    462.       495.       384. </span>
<span class="co">#&gt;  9 ENSG00000004766:VPS50       40.2       38.6      121. </span>
<span class="co">#&gt; 10 ENSG00000004864:SLC25A13   370.       396.       844. </span>
<span class="co">#&gt; # … with 4,024 more rows</span>




<span class="co"># get fold change values for all genes and all comparisons.</span>

<span class="va">q_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000196415:PRTN3"</span>, <span class="st">"ENSG00000221988:PPT2"</span>, <span class="st">"ENSG00000163138:PACRGL"</span>, <span class="st">"ENSG00000183840:GPR39"</span>, <span class="st">"ENSG00000146700:SSC4D"</span>, <span class="st">"ENSG00000163746:PLSCR2"</span>, <span class="st">"ENSG00000155918:RAET1L"</span>, <span class="st">"ENSG00000151458:ANKRD50"</span>, <span class="st">"ENSG00000167074:TEF"</span>, <span class="st">"ENSG00000130159:ECSIT"</span><span class="op">)</span>

<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_fold_change_matrix.html">get_fold_change_matrix</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">comp</span>, 
                                  genes <span class="op">=</span> <span class="va">q_genes</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 10 × 3</span>
<span class="co">#&gt;    gene_id                 treatment1_VS_control treatment2_VS_control</span>
<span class="co">#&gt;    &lt;chr&gt;                                   &lt;dbl&gt;                 &lt;dbl&gt;</span>
<span class="co">#&gt;  1 ENSG00000196415:PRTN3                 -0.0714                 0.729</span>
<span class="co">#&gt;  2 ENSG00000221988:PPT2                   0.0769                 0.699</span>
<span class="co">#&gt;  3 ENSG00000163138:PACRGL                 0.173                  0.691</span>
<span class="co">#&gt;  4 ENSG00000183840:GPR39                  0.170                  0.904</span>
<span class="co">#&gt;  5 ENSG00000146700:SSC4D                  0.0793                 0.773</span>
<span class="co">#&gt;  6 ENSG00000163746:PLSCR2                 0.0321                 0.747</span>
<span class="co">#&gt;  7 ENSG00000155918:RAET1L                 0.467                  2.63 </span>
<span class="co">#&gt;  8 ENSG00000151458:ANKRD50               -0.0441                 1.91 </span>
<span class="co">#&gt;  9 ENSG00000167074:TEF                   -0.0815                 0.681</span>
<span class="co">#&gt; 10 ENSG00000130159:ECSIT                  0.0317                 0.751</span>



<span class="co"># get differentially expressed genes for given comparison </span>

<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_genes_by_regulation.html">get_genes_by_regulation</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                  sample_comparison <span class="op">=</span> <span class="st">"treatment1_VS_control"</span>, 
                                  regulation <span class="op">=</span> <span class="st">"both"</span> <span class="co"># can be one of the "up" , "down" , "both", "other", "all"</span>
                                  <span class="op">)</span>
<span class="co">#&gt;                           up                           up </span>
<span class="co">#&gt;       "ENSG00000187193:MT1X" "ENSG00000269533:AC003002.3" </span>
<span class="co">#&gt;                           up                           up </span>
<span class="co">#&gt;      "ENSG00000065427:KARS1"       "ENSG00000007171:NOS2" </span>
<span class="co">#&gt;                           up                           up </span>
<span class="co">#&gt;     "ENSG00000285447:ZNF883"      "ENSG00000111275:ALDH2" </span>
<span class="co">#&gt;                           up                           up </span>
<span class="co">#&gt;     "ENSG00000279111:OR10X1"   "ENSG00000167617:CDC42EP5" </span>
<span class="co">#&gt;                           up                           up </span>
<span class="co">#&gt;    "ENSG00000186566:GPATCH8"    "ENSG00000232423:PRAMEF6" </span>
<span class="co">#&gt;                           up                           up </span>
<span class="co">#&gt;    "ENSG00000182103:FAM181B"       "ENSG00000172500:FIBP" </span>
<span class="co">#&gt;                           up                         down </span>
<span class="co">#&gt;      "ENSG00000151033:LYZL2"     "ENSG00000250305:TRMT9B" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;      "ENSG00000134308:YWHAQ"    "ENSG00000123575:FAM199X" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;   "ENSG00000148482:SLC39A12"      "ENSG00000175264:CHST1" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;      "ENSG00000141933:TPGS1"      "ENSG00000165621:OXGR1" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;        "ENSG00000104081:BMF"       "ENSG00000278023:RDM1" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;    "ENSG00000172264:MACROD2"       "ENSG00000162971:TYW5" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;     "ENSG00000198682:PAPSS2"      "ENSG00000125445:MRPS7" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;   "ENSG00000138670:RASGEF1B"      "ENSG00000124839:RAB17" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;      "ENSG00000069869:NEDD4"      "ENSG00000174576:NPAS4" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;    "ENSG00000143514:TP53BP2"     "ENSG00000181009:OR52N5" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;      "ENSG00000257315:ZBED6"      "ENSG00000141748:ARL5C" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;     "ENSG00000198886:MT-ND4"      "ENSG00000157045:NTAN1" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;     "ENSG00000165131:LLCFC1"   "ENSG00000277611:Z98752.3" </span>
<span class="co">#&gt;                         down                         down </span>
<span class="co">#&gt;    "ENSG00000158089:GALNT14"      "ENSG00000124784:RIOK1" </span>
<span class="co">#&gt;                         down </span>
<span class="co">#&gt;       "ENSG00000103197:TSC2"</span>

<span class="co"># get replicates group data </span>

<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/group_replicates_by_sample.html">group_replicates_by_sample</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 9 × 2</span>
<span class="co">#&gt;   groups     samples     </span>
<span class="co">#&gt;   &lt;chr&gt;      &lt;chr&gt;       </span>
<span class="co">#&gt; 1 treatment1 treat1_rep1 </span>
<span class="co">#&gt; 2 treatment1 treat1_rep2 </span>
<span class="co">#&gt; 3 treatment1 treat1_rep3 </span>
<span class="co">#&gt; 4 control    control_rep1</span>
<span class="co">#&gt; 5 control    control_rep2</span>
<span class="co">#&gt; 6 control    control_rep3</span>
<span class="co">#&gt; 7 treatment2 treat2_rep1 </span>
<span class="co">#&gt; 8 treatment2 treat2_rep2 </span>
<span class="co">#&gt; 9 treatment2 treat2_rep3</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generate-several-visualizations-from-res">Generate several visualizations from <code>res</code>
<a class="anchor" aria-label="anchor" href="#generate-several-visualizations-from-res"></a>
</h2>
<div class="section level3">
<h3 id="visualize-pairwise-correlation-between-replicates">Visualize pairwise correlation between replicates<a class="anchor" aria-label="anchor" href="#visualize-pairwise-correlation-between-replicates"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_pairwise_corr_plot.html">get_pairwise_corr_plot</a></span><span class="op">(</span><span class="va">res</span>, samples <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Registered S3 method overwritten by 'GGally':</span>
<span class="co">#&gt;   method from   </span>
<span class="co">#&gt;   +.gg   ggplot2</span>
<span class="co">#&gt; $control</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-12-1.png" width="100%"></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $treatment1</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-12-2.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="visualize-samples-by-principle-component-analysis-pca">Visualize samples by Principle Component Analysis (PCA)<a class="anchor" aria-label="anchor" href="#visualize-samples-by-principle-component-analysis-pca"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_pca_plot.html">get_pca_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                        samples  <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span> ,<span class="st">"treatment2"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-13-1.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="visualize-differential-expressed-genes-by-volcano-plot">Visualize differential expressed genes by volcano plot<a class="anchor" aria-label="anchor" href="#visualize-differential-expressed-genes-by-volcano-plot"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_volcano_plot.html">get_volcano_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, sample_comparison <span class="op">=</span> <span class="st">"treatment2_VS_control"</span>,
                            col_up <span class="op">=</span> <span class="st">"#a40000"</span>,
                            col_down <span class="op">=</span> <span class="st">"#007e2f"</span>, 
                            col_other <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span>
<span class="co">#&gt; Registered S3 methods overwritten by 'ggalt':</span>
<span class="co">#&gt;   method                  from   </span>
<span class="co">#&gt;   grid.draw.absoluteGrob  ggplot2</span>
<span class="co">#&gt;   grobHeight.absoluteGrob ggplot2</span>
<span class="co">#&gt;   grobWidth.absoluteGrob  ggplot2</span>
<span class="co">#&gt;   grobX.absoluteGrob      ggplot2</span>
<span class="co">#&gt;   grobY.absoluteGrob      ggplot2</span>
<span class="co">#&gt; Warning: One or more p-values is 0. Converting to 10^-1 * current lowest non-</span>
<span class="co">#&gt; zero p-value...</span>
<span class="co">#&gt; Warning: Ignoring unknown parameters: xlim, ylim</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-14-1.png" width="100%"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># change cutoffs </span>

<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_volcano_plot.html">get_volcano_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                            sample_comparison <span class="op">=</span> <span class="st">"treatment2_VS_control"</span>,
                            pval_cutoff <span class="op">=</span> <span class="fl">0.01</span>,
                            log2fc_cutoff <span class="op">=</span> <span class="fl">0.6</span>, 
                            col_up <span class="op">=</span> <span class="st">"#a40000"</span>,
                            col_down <span class="op">=</span> <span class="st">"#007e2f"</span>,
                            col_other <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span>
<span class="co">#&gt; Warning: One or more p-values is 0. Converting to 10^-1 * current lowest non-</span>
<span class="co">#&gt; zero p-value...</span>

<span class="co">#&gt; Warning: Ignoring unknown parameters: xlim, ylim</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-14-2.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="visualize-gene-expression-distribution-using-box-plot">Visualize gene expression distribution using box plot<a class="anchor" aria-label="anchor" href="#visualize-gene-expression-distribution-using-box-plot"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># all replicates </span>
<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_box_plot.html">get_gene_expression_box_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                        samples <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span><span class="op">)</span>, 
                                        group_replicates <span class="op">=</span> <span class="cn">FALSE</span>,
                                        convert_log2 <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-15-1.png" width="100%"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># summarise  replicates </span>
<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_box_plot.html">get_gene_expression_box_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                        samples <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span><span class="op">)</span>, 
                                        group_replicates <span class="op">=</span> <span class="cn">T</span>,
                                        convert_log2 <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-15-2.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="visualize-genes-by-heatmaps">Visualize genes by heatmaps<a class="anchor" aria-label="anchor" href="#visualize-genes-by-heatmaps"></a>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="va">genes_for_hm</span> <span class="op">=</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_genes_by_regulation.html">get_genes_by_regulation</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>,
                                                  sample_comparison <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">comp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, 
                                                  regulation <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span>

<span class="co"># heatmap of normalised gene expression values across samples </span>

<span class="va">hm1</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_heatmap.html">get_gene_expression_heatmap</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                       samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>,<span class="st">"treatment1"</span> , <span class="st">"treatment2"</span><span class="op">)</span> , 
                                       genes <span class="op">=</span> <span class="va">genes_for_hm</span> , 
                                       convert_zscore <span class="op">=</span> <span class="cn">FALSE</span>, 
                                       convert_log2 <span class="op">=</span> <span class="cn">T</span>, 
                                       summarise_replicates <span class="op">=</span> <span class="cn">T</span>,
                                       name <span class="op">=</span> <span class="st">"log2(value)"</span> , 
                                       cluster_columns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm1</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-16-1.png" width="100%"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Visualise  z-score and show all replicates.</span>

<span class="va">hm2</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_heatmap.html">get_gene_expression_heatmap</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                       samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>,<span class="st">"treatment1"</span><span class="op">)</span> , 
                                       name <span class="op">=</span> <span class="st">"Z-score"</span>,
                                       summarise_replicates <span class="op">=</span> <span class="cn">F</span>, 
                                       genes <span class="op">=</span> <span class="va">genes_for_hm</span> , 
                                       convert_zscore <span class="op">=</span> <span class="cn">TRUE</span>, 
                                       cluster_columns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Warning in parcutils::get_gene_expression_heatmap(x =</span>
<span class="co">#&gt; res, samples = c("control", : Genes having value NA -</span>
<span class="co">#&gt; ENSG00000170820:FSHR,ENSG00000129862:VCY1B are removed from heatmap.</span>


<span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm2</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-16-2.png" width="100%"></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># log2 FC heatamap</span>
<span class="va">hm3</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_heatmap.html">get_fold_change_heatmap</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                   sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">comp</span>, 
                                   genes <span class="op">=</span> <span class="va">genes_for_hm</span> , 
                                   name<span class="op">=</span> <span class="st">"Log2FC"</span><span class="op">)</span>

<span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm3</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-16-3.png" width="100%"></p>
</div>
<div class="section level3">
<h3 id="visualize-differential-genes-overlap-between-comparison">Visualize differential genes overlap between comparison<a class="anchor" aria-label="anchor" href="#visualize-differential-genes-overlap-between-comparison"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">us_plot</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/plot_deg_upsets.html">plot_deg_upsets</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">comp</span><span class="op">)</span>

<span class="va">us_plot</span><span class="op">$</span><span class="va">treatment1_VS_control_AND_treatment2_VS_control</span><span class="op">$</span><span class="va">upset_plot</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-17-1.png" width="100%"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># get list of intersecting genes. </span>

<span class="va">us_plot</span><span class="op">$</span><span class="va">treatment1_VS_control_AND_treatment2_VS_control</span><span class="op">$</span><span class="va">upset_intersects</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 7 × 2</span>
<span class="co">#&gt;   set                                                   elements   </span>
<span class="co">#&gt;   &lt;chr&gt;                                                 &lt;list&gt;     </span>
<span class="co">#&gt; 1 treatment1_VS_control_up                              &lt;chr [6]&gt;  </span>
<span class="co">#&gt; 2 treatment1_VS_control_up,treatment2_VS_control_up     &lt;chr [7]&gt;  </span>
<span class="co">#&gt; 3 treatment2_VS_control_up                              &lt;chr [318]&gt;</span>
<span class="co">#&gt; 4 treatment2_VS_control_up,treatment1_VS_control_down   &lt;chr [8]&gt;  </span>
<span class="co">#&gt; 5 treatment1_VS_control_down,treatment2_VS_control_down &lt;chr [7]&gt;  </span>
<span class="co">#&gt; 6 treatment1_VS_control_down                            &lt;chr [13]&gt; </span>
<span class="co">#&gt; 7 treatment2_VS_control_down                            &lt;chr [495]&gt;</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-genes-by-line-plot">Visualize genes by line plot<a class="anchor" aria-label="anchor" href="#visualize-genes-by-line-plot"></a>
</h3>
<div class="section level4">
<h4 id="section"></h4>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">genes_for_lineplot</span> <span class="op">=</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_genes_by_regulation.html">get_genes_by_regulation</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>,
                                                  sample_comparison <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">comp</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, 
                                                  regulation <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span>

<span class="co">#gene expression values</span>
<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_line_plot.html">get_gene_expression_line_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                   genes <span class="op">=</span> <span class="va">genes_for_lineplot</span> , 
                                   samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>,<span class="st">"treatment1"</span>,<span class="st">"treatment2"</span><span class="op">)</span>,summarise_replicates <span class="op">=</span> <span class="cn">T</span>, show_average_line <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
<span class="co">#&gt; Warning: Transformation introduced infinite values in continuous y-axis</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-18-1.png" width="100%"></p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Fold change values </span>

<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_line_plot.html">get_fold_change_line_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, 
                                   genes <span class="op">=</span> <span class="va">genes_for_lineplot</span> , 
                                     sample_comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1_VS_control"</span>, <span class="st">"treatment2_VS_control"</span><span class="op">)</span>, 
                                   average_line_summary_method <span class="op">=</span>  <span class="st">"mean"</span>,
                                   show_average_line <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-18-2.png" width="100%"></p>
</div>
<div class="section level4">
<h4 id="section-1"></h4>
</div>
</div>
</div>
<div class="section level2">
<h2 id="other-functions">Other functions<a class="anchor" aria-label="anchor" href="#other-functions"></a>
</h2>
<div class="section level3">
<h3 id="alignment-summary">Alignment summary<a class="anchor" aria-label="anchor" href="#alignment-summary"></a>
</h3>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">star_align_log_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span> , <span class="st">"a_Log.final.out"</span> , package <span class="op">=</span> <span class="st">"parcutils"</span><span class="op">)</span>

<span class="va">x</span> <span class="op">=</span>  <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_star_align_log_summary.html">get_star_align_log_summary</a></span><span class="op">(</span>log_file <span class="op">=</span> <span class="va">star_align_log_file</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 13 × 2</span>
<span class="co">#&gt;    type                                    val     </span>
<span class="co">#&gt;    &lt;chr&gt;                                   &lt;chr&gt;   </span>
<span class="co">#&gt;  1 Number of input reads                   41936201</span>
<span class="co">#&gt;  2 Uniquely mapped reads number            40090105</span>
<span class="co">#&gt;  3 Uniquely mapped reads %                 95.60%  </span>
<span class="co">#&gt;  4 Average input read length               300     </span>
<span class="co">#&gt;  5 Average mapped length                   298.24  </span>
<span class="co">#&gt;  6 Number of reads mapped to multiple loci 900879  </span>
<span class="co">#&gt;  7 % of reads mapped to multiple loci      2.15%   </span>
<span class="co">#&gt;  8 Number of reads mapped to too many loci 9335    </span>
<span class="co">#&gt;  9 % of reads mapped to too many loci      0.02%   </span>
<span class="co">#&gt; 10 Number of reads unmapped: too short     921464  </span>
<span class="co">#&gt; 11 % of reads unmapped: too short          2.20%   </span>
<span class="co">#&gt; 12 Number of reads unmapped: other         14418   </span>
<span class="co">#&gt; 13 % of reads unmapped: other              0.03%</span>


<span class="co"># plot alignment summary</span>


<span class="va">star_align_log_file_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span> , package <span class="op">=</span> <span class="st">"parcutils"</span><span class="op">)</span>

<span class="va">star_align_log_files</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="va">star_align_log_file_dir</span>, 
                                   glob <span class="op">=</span> <span class="st">"*Log.final.out"</span> ,
                                   recurse <span class="op">=</span> <span class="cn">T</span>,type <span class="op">=</span> <span class="st">"file"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">star_align_log_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_star_align_log_summary_plot.html">get_star_align_log_summary_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">star_align_log_files</span>,
                                col_total_reads <span class="op">=</span> <span class="st">"red"</span>, 
                                col_mapped_reads  <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> </code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-19-1.png" width="100%"></p>
</div>
</div>
<div class="section level2">
<h2 id="section-2"></h2>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cparsania/parcutils/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/cparsania/parcutils/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing parcutils</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Chirag Parsania <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cparsania/parcutils" class="external-link"><img src="https://img.shields.io/badge/devel%20version-0.1.0-blue.svg"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Chirag Parsania.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
