<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="The package contains utility functions to deal with day to
    day bioinformatics tasks. Most of the functions are related to RNA-seq data analysis and visualization. ">
<title>Utility Functions To Deal Day To Day Bioinformatics Tasks. • parcutils</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Utility Functions To Deal Day To Day Bioinformatics Tasks.">
<meta property="og:description" content="The package contains utility functions to deal with day to
    day bioinformatics tasks. Most of the functions are related to RNA-seq data analysis and visualization. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">parcutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cparsania/parcutils/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="parcutils">parcutils<a class="anchor" aria-label="anchor" href="#parcutils"></a>
</h1></div>
<!-- badges: start -->

<!-- badges: end -->
<p>The goal of <code>parcutils</code> is to provide day to day bioinformatics utility functions. Most of the functions in the package are useful for analyzing and visualizing complex RNA-seq studies.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span><span class="op">)</span> <span class="op">&amp;&amp;</span> <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span> <span class="op">)</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cparsania/parcutils"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"devtools"</span>,<span class="st">"BiocManager"</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span> <span class="op">)</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"cparsania/parcutils"</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="rna-seq-analysis">RNA-seq analysis<a class="anchor" aria-label="anchor" href="#rna-seq-analysis"></a>
</h2>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential expression analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<div class="section level4">
<h4 id="prepare-a-count-table">Prepare a count table<a class="anchor" aria-label="anchor" href="#prepare-a-count-table"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">count_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"toy_counts.txt"</span> , package <span class="op">=</span> <span class="st">"parcutils"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_data</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_delim</a></span><span class="op">(</span><span class="va">count_file</span>, delim <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">count_data</span> </span>
<span><span class="co">#&gt; # A tibble: 5,000 × 10</span></span>
<span><span class="co">#&gt;    gene_id       contr…¹ contr…² contr…³ treat…⁴ treat…⁵ treat…⁶ treat…⁷ treat…⁸</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;           &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 ENSG00000173…       0       0       0       0       0       0       0       0</span></span>
<span><span class="co">#&gt;  2 ENSG00000106…       1       0       2       0       0       1       1       1</span></span>
<span><span class="co">#&gt;  3 ENSG00000131…       2       0       2       0       0       1       1       1</span></span>
<span><span class="co">#&gt;  4 ENSG00000154…     652     690     639     607     453     461     809     994</span></span>
<span><span class="co">#&gt;  5 ENSG00000196…    3372    3631    3188    4644    3168    3514    4789    5466</span></span>
<span><span class="co">#&gt;  6 ENSG00000173…     694     784     829     974     580     716     413     545</span></span>
<span><span class="co">#&gt;  7 ENSG00000140…      87      73      81     100      77      72      63      87</span></span>
<span><span class="co">#&gt;  8 ENSG00000187…       4       4       2       0       5       2       1       4</span></span>
<span><span class="co">#&gt;  9 ENSG00000139…    1374    1789    1564    1933    1459    1584     482     729</span></span>
<span><span class="co">#&gt; 10 ENSG00000165…    3639    4533    3921    3879    3500    3439    3324    4683</span></span>
<span><span class="co">#&gt; # … with 4,990 more rows, 1 more variable: treat2_rep3 &lt;dbl&gt;, and abbreviated</span></span>
<span><span class="co">#&gt; #   variable names ¹​control_rep1, ²​control_rep2, ³​control_rep3, ⁴​treat1_rep1,</span></span>
<span><span class="co">#&gt; #   ⁵​treat1_rep2, ⁶​treat1_rep3, ⁷​treat2_rep1, ⁸​treat2_rep2</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="group-replicates-by-samples">Group replicates by samples<a class="anchor" aria-label="anchor" href="#group-replicates-by-samples"></a>
</h4>
<p>To run DESeq2, replicates for each sample needs to be grouped.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">sample_info</span> <span class="op">&lt;-</span> <span class="va">count_data</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="va">.</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>  <span class="op">%&gt;%</span></span>
<span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="va">.</span> , groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span> , <span class="st">"treatment2"</span><span class="op">)</span> , </span>
<span>                                           each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">sample_info</span></span>
<span><span class="co">#&gt; # A tibble: 9 × 2</span></span>
<span><span class="co">#&gt;   samples      groups    </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;        &lt;chr&gt;     </span></span>
<span><span class="co">#&gt; 1 control_rep1 control   </span></span>
<span><span class="co">#&gt; 2 control_rep2 control   </span></span>
<span><span class="co">#&gt; 3 control_rep3 control   </span></span>
<span><span class="co">#&gt; 4 treat1_rep1  treatment1</span></span>
<span><span class="co">#&gt; 5 treat1_rep2  treatment1</span></span>
<span><span class="co">#&gt; 6 treat1_rep3  treatment1</span></span>
<span><span class="co">#&gt; 7 treat2_rep1  treatment2</span></span>
<span><span class="co">#&gt; 8 treat2_rep2  treatment2</span></span>
<span><span class="co">#&gt; 9 treat2_rep3  treatment2</span></span></code></pre></div>
<blockquote>
<p>NOTE: Samples which are present in the object ‘sample_info’ will be considered for differential expressed analysis.</p>
</blockquote>
</div>
<div class="section level4">
<h4 id="run-deseq2-for-multiple-differential-gene-comparison">Run <code>DESeq2</code> for multiple differential gene comparison.<a class="anchor" aria-label="anchor" href="#run-deseq2-for-multiple-differential-gene-comparison"></a>
</h4>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/run_deseq_analysis.html">run_deseq_analysis</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_data</span> ,</span>
<span>                         sample_info <span class="op">=</span> <span class="va">sample_info</span>,</span>
<span>                         column_geneid <span class="op">=</span> <span class="st">"gene_id"</span> ,</span>
<span>                         cutoff_lfc <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                         cutoff_pval <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                         group_numerator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1"</span>, <span class="st">"treatment2"</span><span class="op">)</span> ,</span>
<span>                         group_denominator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="lets-have-a-look-in-to-res">Let’s have a look in to <code>res</code>
<a class="anchor" aria-label="anchor" href="#lets-have-a-look-in-to-res"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">res</span></span>
<span><span class="co">#&gt; ┌────────────────────────────┐</span></span>
<span><span class="co">#&gt; │                            │</span></span>
<span><span class="co">#&gt; │   Summary of DE analysis   │</span></span>
<span><span class="co">#&gt; │                            │</span></span>
<span><span class="co">#&gt; └────────────────────────────┘</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; treatment1_VS_control </span></span>
<span><span class="co">#&gt; • number of up genes   : 13.</span></span>
<span><span class="co">#&gt; • number of down genes : 28.</span></span>
<span><span class="co">#&gt; ──────────────────────────────</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; treatment2_VS_control </span></span>
<span><span class="co">#&gt; • number of up genes   : 333.</span></span>
<span><span class="co">#&gt; • number of down genes : 502.</span></span>
<span><span class="co">#&gt; ──────────────────────────────</span></span></code></pre></div>
<p><code>res</code> is an object of improved <code>dataframe</code> - <code>tibble</code>. Each row in the <code>res</code> is a differential comparison which can be identified by the value from the column <code>comp</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span></span>
<span><span class="co">#&gt; [1] "treatment1_VS_control" "treatment2_VS_control"</span></span></code></pre></div>
<p>Data related to each differential comparison can be found from other columns of <code>res</code>.</p>
<p>For example, summary of differently expressed genes can be found from the column <code>deg_summmary</code></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">deg_summmary</span></span>
<span><span class="co">#&gt; $treatment1_VS_control</span></span>
<span><span class="co">#&gt; # A tibble: 3 × 2</span></span>
<span><span class="co">#&gt;   regul     n</span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 Down     28</span></span>
<span><span class="co">#&gt; 2 other  3993</span></span>
<span><span class="co">#&gt; 3 Up       13</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $treatment2_VS_control</span></span>
<span><span class="co">#&gt; # A tibble: 3 × 2</span></span>
<span><span class="co">#&gt;   regul     n</span></span>
<span><span class="co">#&gt;   &lt;chr&gt; &lt;int&gt;</span></span>
<span><span class="co">#&gt; 1 Down    502</span></span>
<span><span class="co">#&gt; 2 other  3199</span></span>
<span><span class="co">#&gt; 3 Up      333</span></span></code></pre></div>
<p>As described below there are several helper functions to get data from the <code>res</code> .</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="get-data-from-res-using-helper-functions">Get data from <code>res</code> using helper functions<a class="anchor" aria-label="anchor" href="#get-data-from-res-using-helper-functions"></a>
</h2>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># get normalised gene expression value for all genes across all samples. </span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_normalised_expression_matrix.html">get_normalised_expression_matrix</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                            samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                            genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                            summarise_replicates <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 4,034 × 10</span></span>
<span><span class="co">#&gt;    gene_id       treat…¹ treat…² treat…³ contr…⁴ contr…⁵ contr…⁶ treat…⁷ treat…⁸</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;           &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;   &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 ENSG00000154…  4.40e2  4.61e2 4.56e+2   694.   6.87e2  6.76e2  8.76e2  1.07e3</span></span>
<span><span class="co">#&gt;  2 ENSG00000196…  3.37e3  3.22e3 3.48e+3  3591.   3.62e3  3.37e3  5.19e3  5.87e3</span></span>
<span><span class="co">#&gt;  3 ENSG00000173…  7.07e2  5.90e2 7.09e+2   739.   7.81e2  8.77e2  4.47e2  5.85e2</span></span>
<span><span class="co">#&gt;  4 ENSG00000140…  7.26e1  7.83e1 7.13e+1    92.7  7.27e1  8.57e1  6.82e1  9.34e1</span></span>
<span><span class="co">#&gt;  5 ENSG00000139…  1.40e3  1.48e3 1.57e+3  1463.   1.78e3  1.66e3  5.22e2  7.82e2</span></span>
<span><span class="co">#&gt;  6 ENSG00000165…  2.81e3  3.56e3 3.40e+3  3875.   4.51e3  4.15e3  3.60e3  5.03e3</span></span>
<span><span class="co">#&gt;  7 ENSG00000179…  1.27e4  1.25e4 1.37e+4 13257.   1.34e4  1.40e4  1.08e4  1.51e4</span></span>
<span><span class="co">#&gt;  8 ENSG00000188…  4.61e2  4.34e2 4.59e+2   540.   5.92e2  5.66e2  4.68e2  5.80e2</span></span>
<span><span class="co">#&gt;  9 ENSG00000249…  0       2.03e0 9.90e-1     0    1.10e1  2.12e0  2.17e0  1.07e0</span></span>
<span><span class="co">#&gt; 10 ENSG00000110…  2.18e0  9.15e0 9.90e+0    11.7  5.97e0  1.59e1  5.42e0  4.29e0</span></span>
<span><span class="co">#&gt; # … with 4,024 more rows, 1 more variable: treat2_rep3 &lt;dbl&gt;, and abbreviated</span></span>
<span><span class="co">#&gt; #   variable names ¹​treat1_rep1, ²​treat1_rep2, ³​treat1_rep3, ⁴​control_rep1,</span></span>
<span><span class="co">#&gt; #   ⁵​control_rep2, ⁶​control_rep3, ⁷​treat2_rep1, ⁸​treat2_rep2</span></span>
<span></span>
<span><span class="co"># average gene expression values across relicates  </span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_normalised_expression_matrix.html">get_normalised_expression_matrix</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                            samples <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                            genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                                            summarise_replicates <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                                            summarise_method <span class="op">=</span> <span class="st">"median"</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 4,034 × 4</span></span>
<span><span class="co">#&gt;    gene_id                  control treatment1 treatment2</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                      &lt;dbl&gt;      &lt;dbl&gt;      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 ENSG00000000971:CFH       7349.      5727.      9849. </span></span>
<span><span class="co">#&gt;  2 ENSG00000001461:NIPAL3     134.       126.        68.2</span></span>
<span><span class="co">#&gt;  3 ENSG00000001497:LAS1L     9282.      8228.      7678. </span></span>
<span><span class="co">#&gt;  4 ENSG00000001631:KRIT1    23294.     29183.     20375. </span></span>
<span><span class="co">#&gt;  5 ENSG00000002746:HECW1       66.0       31.9       24.9</span></span>
<span><span class="co">#&gt;  6 ENSG00000003056:M6PR      6538.      7394.      6694. </span></span>
<span><span class="co">#&gt;  7 ENSG00000003436:TFPI       248.       225.       200. </span></span>
<span><span class="co">#&gt;  8 ENSG00000003509:NDUFAF7    462.       495.       384. </span></span>
<span><span class="co">#&gt;  9 ENSG00000004766:VPS50       40.2       38.6      121. </span></span>
<span><span class="co">#&gt; 10 ENSG00000004864:SLC25A13   370.       396.       844. </span></span>
<span><span class="co">#&gt; # … with 4,024 more rows</span></span>
<span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># get fold change values for all genes and all comparisons.</span></span>
<span></span>
<span><span class="va">q_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ENSG00000196415:PRTN3"</span>, <span class="st">"ENSG00000221988:PPT2"</span>, <span class="st">"ENSG00000163138:PACRGL"</span>, <span class="st">"ENSG00000183840:GPR39"</span>, <span class="st">"ENSG00000146700:SSC4D"</span>, <span class="st">"ENSG00000163746:PLSCR2"</span>, <span class="st">"ENSG00000155918:RAET1L"</span>, <span class="st">"ENSG00000151458:ANKRD50"</span>, <span class="st">"ENSG00000167074:TEF"</span>, <span class="st">"ENSG00000130159:ECSIT"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_fold_change_matrix.html">get_fold_change_matrix</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                  sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span>, </span>
<span>                                  genes <span class="op">=</span> <span class="va">q_genes</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 10 × 3</span></span>
<span><span class="co">#&gt;    gene_id                 treatment1_VS_control treatment2_VS_control</span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                                   &lt;dbl&gt;                 &lt;dbl&gt;</span></span>
<span><span class="co">#&gt;  1 ENSG00000196415:PRTN3                 -0.0714                 0.729</span></span>
<span><span class="co">#&gt;  2 ENSG00000221988:PPT2                   0.0769                 0.699</span></span>
<span><span class="co">#&gt;  3 ENSG00000163138:PACRGL                 0.173                  0.691</span></span>
<span><span class="co">#&gt;  4 ENSG00000183840:GPR39                  0.170                  0.904</span></span>
<span><span class="co">#&gt;  5 ENSG00000146700:SSC4D                  0.0793                 0.773</span></span>
<span><span class="co">#&gt;  6 ENSG00000163746:PLSCR2                 0.0321                 0.747</span></span>
<span><span class="co">#&gt;  7 ENSG00000155918:RAET1L                 0.467                  2.63 </span></span>
<span><span class="co">#&gt;  8 ENSG00000151458:ANKRD50               -0.0441                 1.91 </span></span>
<span><span class="co">#&gt;  9 ENSG00000167074:TEF                   -0.0815                 0.681</span></span>
<span><span class="co">#&gt; 10 ENSG00000130159:ECSIT                  0.0317                 0.751</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># get differentially expressed genes for given comparison </span></span>
<span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_genes_by_regulation.html">get_genes_by_regulation</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                  sample_comparison <span class="op">=</span> <span class="st">"treatment1_VS_control"</span>, </span>
<span>                                  regulation <span class="op">=</span> <span class="st">"both"</span> <span class="co"># can be one of the "up" , "down" , "both", "other", "all"</span></span>
<span>                                  <span class="op">)</span></span>
<span><span class="co">#&gt;     ENSG00000250305:TRMT9B      ENSG00000134308:YWHAQ </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;    ENSG00000123575:FAM199X   ENSG00000148482:SLC39A12 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;      ENSG00000175264:CHST1      ENSG00000141933:TPGS1 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;      ENSG00000165621:OXGR1        ENSG00000104081:BMF </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;       ENSG00000278023:RDM1    ENSG00000172264:MACROD2 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;       ENSG00000162971:TYW5     ENSG00000198682:PAPSS2 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;      ENSG00000125445:MRPS7   ENSG00000138670:RASGEF1B </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;      ENSG00000124839:RAB17      ENSG00000069869:NEDD4 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;      ENSG00000174576:NPAS4    ENSG00000143514:TP53BP2 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;     ENSG00000181009:OR52N5      ENSG00000257315:ZBED6 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;      ENSG00000141748:ARL5C     ENSG00000198886:MT-ND4 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;      ENSG00000157045:NTAN1     ENSG00000165131:LLCFC1 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;   ENSG00000277611:Z98752.3    ENSG00000158089:GALNT14 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;      ENSG00000124784:RIOK1       ENSG00000103197:TSC2 </span></span>
<span><span class="co">#&gt;                       down                       down </span></span>
<span><span class="co">#&gt;       ENSG00000187193:MT1X ENSG00000269533:AC003002.3 </span></span>
<span><span class="co">#&gt;                         up                         up </span></span>
<span><span class="co">#&gt;      ENSG00000065427:KARS1       ENSG00000007171:NOS2 </span></span>
<span><span class="co">#&gt;                         up                         up </span></span>
<span><span class="co">#&gt;     ENSG00000285447:ZNF883      ENSG00000111275:ALDH2 </span></span>
<span><span class="co">#&gt;                         up                         up </span></span>
<span><span class="co">#&gt;     ENSG00000279111:OR10X1   ENSG00000167617:CDC42EP5 </span></span>
<span><span class="co">#&gt;                         up                         up </span></span>
<span><span class="co">#&gt;    ENSG00000186566:GPATCH8    ENSG00000232423:PRAMEF6 </span></span>
<span><span class="co">#&gt;                         up                         up </span></span>
<span><span class="co">#&gt;    ENSG00000182103:FAM181B       ENSG00000172500:FIBP </span></span>
<span><span class="co">#&gt;                         up                         up </span></span>
<span><span class="co">#&gt;      ENSG00000151033:LYZL2 </span></span>
<span><span class="co">#&gt;                         up </span></span>
<span><span class="co">#&gt; Levels: up down other</span></span>
<span></span>
<span><span class="co"># get replicates group data </span></span>
<span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/dot-group_replicates_by_sample.html">.group_replicates_by_sample</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 9 × 2</span></span>
<span><span class="co">#&gt;   groups     samples     </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;      &lt;chr&gt;       </span></span>
<span><span class="co">#&gt; 1 treatment1 treat1_rep1 </span></span>
<span><span class="co">#&gt; 2 treatment1 treat1_rep2 </span></span>
<span><span class="co">#&gt; 3 treatment1 treat1_rep3 </span></span>
<span><span class="co">#&gt; 4 control    control_rep1</span></span>
<span><span class="co">#&gt; 5 control    control_rep2</span></span>
<span><span class="co">#&gt; 6 control    control_rep3</span></span>
<span><span class="co">#&gt; 7 treatment2 treat2_rep1 </span></span>
<span><span class="co">#&gt; 8 treatment2 treat2_rep2 </span></span>
<span><span class="co">#&gt; 9 treatment2 treat2_rep3</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="generate-several-visualizations-from-res">Generate several visualizations from <code>res</code>
<a class="anchor" aria-label="anchor" href="#generate-several-visualizations-from-res"></a>
</h2>
<div class="section level3">
<h3 id="visualize-pairwise-correlation-between-replicates">Visualize pairwise correlation between replicates<a class="anchor" aria-label="anchor" href="#visualize-pairwise-correlation-between-replicates"></a>
</h3>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_pairwise_corr_plot.html">get_pairwise_corr_plot</a></span><span class="op">(</span><span class="va">res</span>, samples <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; $control</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-12-1.png"><!-- --></p>
<div class="sourceCode" id="cb10"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="co">#&gt; $treatment1</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-12-2.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="visualize-all-sample-correlation-by-heat-box">Visualize all sample correlation by heat box<a class="anchor" aria-label="anchor" href="#visualize-all-sample-correlation-by-heat-box"></a>
</h3>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_corr_heatbox.html">get_corr_heatbox</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, show_corr_values <span class="op">=</span> <span class="cn">T</span>, cluster_samples <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-13-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="visualize-samples-by-principle-component-analysis-pca">Visualize samples by Principle Component Analysis (PCA)<a class="anchor" aria-label="anchor" href="#visualize-samples-by-principle-component-analysis-pca"></a>
</h3>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_pca_plot.html">get_pca_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                        samples  <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span> ,<span class="st">"treatment2"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-14-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="counts-of-diff-expressed-genes">Counts of diff expressed genes<a class="anchor" aria-label="anchor" href="#counts-of-diff-expressed-genes"></a>
</h3>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_diff_gene_count_barplot.html">get_diff_gene_count_barplot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-15-1.png"><!-- --></p>
<p>change color of the bars</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_diff_gene_count_barplot.html">get_diff_gene_count_barplot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, col_down <span class="op">=</span> <span class="st">"green4"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-16-1.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="visualize-differential-expressed-genes-by-volcano-plot">Visualize differential expressed genes by volcano plot<a class="anchor" aria-label="anchor" href="#visualize-differential-expressed-genes-by-volcano-plot"></a>
</h3>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_volcano_plot.html">get_volcano_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, sample_comparison <span class="op">=</span> <span class="st">"treatment2_VS_control"</span>,</span>
<span>                            col_up <span class="op">=</span> <span class="st">"#a40000"</span>,</span>
<span>                            col_down <span class="op">=</span> <span class="st">"#16317d"</span>, </span>
<span>                            repair_genes <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                            col_other <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-17-1.png"><!-- --></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># change cutoffs </span></span>
<span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_volcano_plot.html">get_volcano_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, repair_genes <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                            sample_comparison <span class="op">=</span> <span class="st">"treatment2_VS_control"</span>,</span>
<span>                            pval_cutoff <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                            log2fc_cutoff <span class="op">=</span> <span class="fl">0.6</span>, </span>
<span>                            col_up <span class="op">=</span> <span class="st">"#a40000"</span>,</span>
<span>                            col_down <span class="op">=</span> <span class="st">"#16317d"</span>,</span>
<span>                            col_other <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-17-2.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="visualize-gene-expression-distribution-using-box-plot">Visualize gene expression distribution using box plot<a class="anchor" aria-label="anchor" href="#visualize-gene-expression-distribution-using-box-plot"></a>
</h3>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># all replicates </span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_box_plot.html">get_gene_expression_box_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                        samples <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span><span class="op">)</span>, </span>
<span>                                        group_replicates <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                        convert_log2 <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-18-1.png"><!-- --></p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># summarise  replicates </span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_box_plot.html">get_gene_expression_box_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                        samples <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span><span class="op">)</span>, </span>
<span>                                        group_replicates <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                                        convert_log2 <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-18-2.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="visualize-genes-by-heatmaps">Visualize genes by heatmaps<a class="anchor" aria-label="anchor" href="#visualize-genes-by-heatmaps"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span></span>
<span><span class="va">genes_for_hm</span> <span class="op">=</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_genes_by_regulation.html">get_genes_by_regulation</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>,</span>
<span>                                                  sample_comparison <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                                  regulation <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># heatmap of normalised gene expression values across samples </span></span>
<span></span>
<span><span class="va">hm1</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_heatmap.html">get_gene_expression_heatmap</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                       samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>,<span class="st">"treatment1"</span> , <span class="st">"treatment2"</span><span class="op">)</span> , </span>
<span>                                       genes <span class="op">=</span> <span class="va">genes_for_hm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span> , </span>
<span>                                       convert_zscore <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                       convert_log2 <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                                       summarise_replicates <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                                       name <span class="op">=</span> <span class="st">"log2(value)"</span> , color_default <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                                       col <span class="op">=</span> </span>
<span>                                         <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>,<span class="fl">0</span>,<span class="fl">15</span><span class="op">)</span>, colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#16317d"</span>,<span class="st">"white"</span>,<span class="st">"#a40000"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                       cluster_columns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm1</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-19-1.png"><!-- --></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Visualise  z-score and show all replicates.</span></span>
<span></span>
<span><span class="va">hm2</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_heatmap.html">get_gene_expression_heatmap</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                       samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>,<span class="st">"treatment1"</span><span class="op">)</span> , </span>
<span>                                       name <span class="op">=</span> <span class="st">"Z-score"</span>,</span>
<span>                                       summarise_replicates <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                                        col <span class="op">=</span> </span>
<span>                                         <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2</span>,<span class="fl">0</span>,<span class="fl">2</span><span class="op">)</span>, colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#16317d"</span>,<span class="st">"white"</span>,<span class="st">"#a40000"</span><span class="op">)</span><span class="op">)</span>,color_default <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                                       genes <span class="op">=</span> <span class="va">genes_for_hm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span> , </span>
<span>                                       convert_zscore <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                       cluster_columns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm2</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-19-2.png"><!-- --></p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># log2 FC heatamap</span></span>
<span><span class="va">hm3</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_heatmap.html">get_fold_change_heatmap</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                   sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span>, </span>
<span>                                   genes <span class="op">=</span> <span class="va">genes_for_hm</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span> , </span>
<span>                                   color_default <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                                     col <span class="op">=</span> </span>
<span>                                         <span class="fu">circlize</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">colorRamp2</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">5</span>,<span class="fl">0</span>,<span class="fl">5</span><span class="op">)</span>, colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#16317d"</span>,<span class="st">"white"</span>,<span class="st">"#a40000"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                   name<span class="op">=</span> <span class="st">"Log2FC"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">hm3</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-19-3.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="visualize-differential-genes-overlap-between-comparison">Visualize differential genes overlap between comparison<a class="anchor" aria-label="anchor" href="#visualize-differential-genes-overlap-between-comparison"></a>
</h3>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">us_plot</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/plot_deg_upsets.html">plot_deg_upsets</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                      sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span><span class="op">)</span></span>
<span></span>
<span><span class="va">us_plot</span><span class="op">$</span><span class="va">treatment1_VS_control_AND_treatment2_VS_control</span><span class="op">$</span><span class="va">upset_plot</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-20-1.png"><!-- --></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># get list of intersecting genes. </span></span>
<span></span>
<span><span class="va">us_plot</span><span class="op">$</span><span class="va">treatment1_VS_control_AND_treatment2_VS_control</span><span class="op">$</span><span class="va">upset_intersects</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 7 × 2</span></span>
<span><span class="co">#&gt;   set                                                   elements   </span></span>
<span><span class="co">#&gt;   &lt;chr&gt;                                                 &lt;list&gt;     </span></span>
<span><span class="co">#&gt; 1 treatment1_VS_control_up                              &lt;chr [6]&gt;  </span></span>
<span><span class="co">#&gt; 2 treatment1_VS_control_up,treatment2_VS_control_up     &lt;chr [7]&gt;  </span></span>
<span><span class="co">#&gt; 3 treatment2_VS_control_up                              &lt;chr [318]&gt;</span></span>
<span><span class="co">#&gt; 4 treatment2_VS_control_up,treatment1_VS_control_down   &lt;chr [8]&gt;  </span></span>
<span><span class="co">#&gt; 5 treatment1_VS_control_down,treatment2_VS_control_down &lt;chr [7]&gt;  </span></span>
<span><span class="co">#&gt; 6 treatment1_VS_control_down                            &lt;chr [13]&gt; </span></span>
<span><span class="co">#&gt; 7 treatment2_VS_control_down                            &lt;chr [495]&gt;</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualize-common-de-genes-between-comparison-by-scatter-plot">Visualize common DE genes between comparison by scatter plot<a class="anchor" aria-label="anchor" href="#visualize-common-de-genes-between-comparison-by-scatter-plot"></a>
</h3>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># show common up and down genes </span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_fold_change_scatter_plot.html">get_fold_change_scatter_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                                   sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span>, point_size <span class="op">=</span> <span class="fl">3</span>,label_size <span class="op">=</span> <span class="fl">3</span>,repair_genes <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-21-1.png"><!-- --></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># show common up and down genes </span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_fold_change_scatter_plot.html">get_fold_change_scatter_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                        sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span>, </span>
<span>                                        point_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                                        label_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                                        repair_genes <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-21-2.png"><!-- --></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># show common up genes</span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_fold_change_scatter_plot.html">get_fold_change_scatter_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                        sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span>, </span>
<span>                                        point_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                        label_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                                        color_label <span class="op">=</span> <span class="st">"both_up"</span>,</span>
<span>                                        col_up <span class="op">=</span> <span class="st">"red"</span>,</span>
<span>                                        repair_genes <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-21-3.png"><!-- --></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="co"># show common down genes</span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_fold_change_scatter_plot.html">get_fold_change_scatter_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                        sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span>, </span>
<span>                                        point_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                        label_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                                        color_label <span class="op">=</span> <span class="st">"both_down"</span>,</span>
<span>                                        col_down <span class="op">=</span> <span class="st">"green4"</span>,</span>
<span>                                        repair_genes <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-21-4.png"><!-- --></p>
</div>
<div class="section level3">
<h3 id="visualize-genes-by-line-plot">Visualize genes by line plot<a class="anchor" aria-label="anchor" href="#visualize-genes-by-line-plot"></a>
</h3>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="va">genes_for_lineplot</span> <span class="op">=</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_genes_by_regulation.html">get_genes_by_regulation</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>,</span>
<span>                                                  sample_comparison <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, </span>
<span>                                                  regulation <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># line plot of gene expression values</span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_line_plot.html">get_gene_expression_line_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                   genes <span class="op">=</span> <span class="va">genes_for_lineplot</span> , </span>
<span>                                   samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>,<span class="st">"treatment1"</span>,<span class="st">"treatment2"</span><span class="op">)</span>,summarise_replicates <span class="op">=</span> <span class="cn">T</span>, show_average_line <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-1.png"><!-- --></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span></span>
<span><span class="co"># line plot of gene expression values with k-means clustering  </span></span>
<span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_line_plot.html">get_gene_expression_line_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                         km <span class="op">=</span> <span class="fl">4</span>,</span>
<span>                                   genes <span class="op">=</span> <span class="va">genes_for_lineplot</span> , </span>
<span>                                   samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>,<span class="st">"treatment1"</span>,<span class="st">"treatment2"</span><span class="op">)</span>,summarise_replicates <span class="op">=</span> <span class="cn">T</span>, show_average_line <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-2.png"><!-- --></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># line plot of gene expression values with k-means clustering  </span></span>
<span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_line_plot.html">get_gene_expression_line_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                         km <span class="op">=</span> <span class="fl">4</span>, </span>
<span>                                         facet_clusters <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                                   genes <span class="op">=</span> <span class="va">genes_for_lineplot</span> , </span>
<span>                                   samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>,<span class="st">"treatment1"</span>,<span class="st">"treatment2"</span><span class="op">)</span>,summarise_replicates <span class="op">=</span> <span class="cn">T</span>, show_average_line <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>, </span>
<span>                 axis.text.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">40</span>,hjust <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-3.png"><!-- --></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Fold change values </span></span>
<span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_gene_expression_line_plot.html">get_fold_change_line_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span>, </span>
<span>                                   genes <span class="op">=</span> <span class="va">genes_for_lineplot</span> , </span>
<span>                                   line_transparency <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                                   km <span class="op">=</span> <span class="fl">2</span>,facet_clusters <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                                     sample_comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1_VS_control"</span>, <span class="st">"treatment2_VS_control"</span><span class="op">)</span>, </span>
<span>                                   average_line_summary_method <span class="op">=</span>  <span class="st">"mean"</span>,</span>
<span>                                   show_average_line <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">15</span><span class="op">)</span>,</span>
<span>                 axis.text.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">40</span>,hjust <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-22-4.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="perform-gene-ontology-analysis-and-visualization-of-all-updown-genes-from-all-comparisons-in-one-go">Perform gene ontology analysis and visualization of all UP/DOWN genes from all comparisons in one go.<a class="anchor" aria-label="anchor" href="#perform-gene-ontology-analysis-and-visualization-of-all-updown-genes-from-all-comparisons-in-one-go"></a>
</h2>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_results</span> <span class="op">&lt;-</span> <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_go_emap_plot.html">get_go_emap_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># GO results as a table </span></span>
<span></span>
<span><span class="va">go_results</span><span class="op">$</span><span class="va">go_enrichment_output</span></span>
<span><span class="co">#&gt; list()</span></span>
<span></span>
<span><span class="co"># GO results as an emap plot </span></span>
<span></span>
<span><span class="va">go_results</span><span class="op">$</span><span class="va">go_emap_plots</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="show-overlapping-genes-through-venn-diagram">Show overlapping genes through VENN diagram<a class="anchor" aria-label="anchor" href="#show-overlapping-genes-through-venn-diagram"></a>
</h2>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/plot_deg_venn.html">plot_deg_venn</a></span><span class="op">(</span><span class="va">res</span>, sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span>,regulation <span class="op">=</span> <span class="st">"up"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-24-1.png"><!-- --></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/plot_deg_venn.html">plot_deg_venn</a></span><span class="op">(</span><span class="va">res</span>, sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span>,regulation <span class="op">=</span> <span class="st">"down"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-24-2.png"><!-- --></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/plot_deg_venn.html">plot_deg_venn</a></span><span class="op">(</span><span class="va">res</span>, sample_comparisons <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">de_comparisons</span>,regulation <span class="op">=</span> <span class="st">"both"</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-24-3.png"><!-- --></p>
</div>
<div class="section level2">
<h2 id="other-functions">Other functions<a class="anchor" aria-label="anchor" href="#other-functions"></a>
</h2>
<div class="section level3">
<h3 id="alignment-summary">Alignment summary<a class="anchor" aria-label="anchor" href="#alignment-summary"></a>
</h3>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">star_align_log_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span> , <span class="st">"a_Log.final.out"</span> , package <span class="op">=</span> <span class="st">"parcutils"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span> <span class="op">=</span>  <span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_star_align_log_summary.html">get_star_align_log_summary</a></span><span class="op">(</span>log_file <span class="op">=</span> <span class="va">star_align_log_file</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 13 × 2</span></span>
<span><span class="co">#&gt;    type                                    val     </span></span>
<span><span class="co">#&gt;    &lt;chr&gt;                                   &lt;chr&gt;   </span></span>
<span><span class="co">#&gt;  1 Number of input reads                   41936201</span></span>
<span><span class="co">#&gt;  2 Uniquely mapped reads number            40090105</span></span>
<span><span class="co">#&gt;  3 Uniquely mapped reads %                 95.60%  </span></span>
<span><span class="co">#&gt;  4 Average input read length               300     </span></span>
<span><span class="co">#&gt;  5 Average mapped length                   298.24  </span></span>
<span><span class="co">#&gt;  6 Number of reads mapped to multiple loci 900879  </span></span>
<span><span class="co">#&gt;  7 % of reads mapped to multiple loci      2.15%   </span></span>
<span><span class="co">#&gt;  8 Number of reads mapped to too many loci 9335    </span></span>
<span><span class="co">#&gt;  9 % of reads mapped to too many loci      0.02%   </span></span>
<span><span class="co">#&gt; 10 Number of reads unmapped: too short     921464  </span></span>
<span><span class="co">#&gt; 11 % of reads unmapped: too short          2.20%   </span></span>
<span><span class="co">#&gt; 12 Number of reads unmapped: other         14418   </span></span>
<span><span class="co">#&gt; 13 % of reads unmapped: other              0.03%</span></span>
<span></span>
<span></span>
<span><span class="co"># plot alignment summary</span></span>
<span></span>
<span></span>
<span><span class="va">star_align_log_file_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span> , package <span class="op">=</span> <span class="st">"parcutils"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">star_align_log_files</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="va">star_align_log_file_dir</span>, </span>
<span>                                   glob <span class="op">=</span> <span class="st">"*Log.final.out"</span> ,</span>
<span>                                   recurse <span class="op">=</span> <span class="cn">T</span>,type <span class="op">=</span> <span class="st">"file"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">star_align_log_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu">parcutils</span><span class="fu">::</span><span class="fu"><a href="reference/get_star_align_log_summary_plot.html">get_star_align_log_summary_plot</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">star_align_log_files</span>,</span>
<span>                                col_total_reads <span class="op">=</span> <span class="st">"red"</span>, </span>
<span>                                col_mapped_reads  <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-25-1.png"><!-- --></p>
</div>
</div>
<div class="section level2">
<h2 id="section"></h2>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cparsania/parcutils/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/cparsania/parcutils/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing parcutils</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Chirag Parsania <br><small class="roles"> Maintainer </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/cparsania/parcutils" class="external-link"><img src="https://img.shields.io/badge/devel%20version-0.1.0-blue.svg"></a></li>
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Chirag Parsania.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
