<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Heatmap is a common tool to show gene expression pattern across samples in RNA-seq experiment.
While doing data exploration, it is a common practice to generate several heatmaps to identify interesting
patterns of gene expression across sample. Most heatmap generating tools require data in a tabular form.
However, to prepare such a tabular form requires significant data wrangling such as sub-setting
genes (rows) and samples (columns) specifically when RNA-seq studies studies consisting of several samples and
replicates.
This function cut downs several steps of data wrangling to create a heatmap of gene expression, z-score or
log2 fold-change . Given an object of 'parcutils', names of samples or sample comprisons, genes to show in heatmap
and several other arguments it creates a heatmap.The output of the function is an output of the function ComplexHeatmap::Heatmap()
which then can be used for the other functions of the ComplexHeatmap package."><title>Generate a heatmap of normalised gene expression values, z-score or log2Fold-change values. — get_gene_expression_heatmap • parcutils</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate a heatmap of normalised gene expression values, z-score or log2Fold-change values. — get_gene_expression_heatmap"><meta property="og:description" content="Heatmap is a common tool to show gene expression pattern across samples in RNA-seq experiment.
While doing data exploration, it is a common practice to generate several heatmaps to identify interesting
patterns of gene expression across sample. Most heatmap generating tools require data in a tabular form.
However, to prepare such a tabular form requires significant data wrangling such as sub-setting
genes (rows) and samples (columns) specifically when RNA-seq studies studies consisting of several samples and
replicates.
This function cut downs several steps of data wrangling to create a heatmap of gene expression, z-score or
log2 fold-change . Given an object of 'parcutils', names of samples or sample comprisons, genes to show in heatmap
and several other arguments it creates a heatmap.The output of the function is an output of the function ComplexHeatmap::Heatmap()
which then can be used for the other functions of the ComplexHeatmap package."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">parcutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cparsania/parcutils/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Generate a heatmap of normalised gene expression values, z-score or log2Fold-change values.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cparsania/parcutils/blob/HEAD/R/rnaseq_related.R" class="external-link"><code>R/rnaseq_related.R</code></a></small>
      <div class="d-none name"><code>get_gene_expression_heatmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Heatmap is a common tool to show gene expression pattern across samples in RNA-seq experiment.
While doing data exploration, it is a common practice to generate several heatmaps to identify interesting
patterns of gene expression across sample. Most heatmap generating tools require data in a tabular form.
However, to prepare such a tabular form requires significant data wrangling such as sub-setting
genes (rows) and samples (columns) specifically when RNA-seq studies studies consisting of several samples and
replicates.</p>
<p>This function cut downs several steps of data wrangling to create a heatmap of gene expression, z-score or
log2 fold-change . Given an object of 'parcutils', names of samples or sample comprisons, genes to show in heatmap
and several other arguments it creates a heatmap.The output of the function is an output of the function <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>
which then can be used for the other functions of the <a href="https://jokergoo.github.io/ComplexHeatmap-reference/book/" class="external-link">ComplexHeatmap</a> package.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">get_gene_expression_heatmap</span><span class="op">(</span>
  <span class="va">x</span>,
  <span class="va">samples</span>,
  <span class="va">genes</span>,
  repair_genes <span class="op">=</span> <span class="cn">FALSE</span>,
  convert_log2 <span class="op">=</span> <span class="cn">FALSE</span>,
  color_default <span class="op">=</span> <span class="cn">TRUE</span>,
  col <span class="op">=</span> <span class="cn">NULL</span>,
  convert_zscore <span class="op">=</span> <span class="cn">TRUE</span>,
  summarise_replicates <span class="op">=</span> <span class="cn">TRUE</span>,
  summarise_method <span class="op">=</span> <span class="st">"median"</span>,
  show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,
  cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>,
  show_row_dend <span class="op">=</span> <span class="cn">TRUE</span>,
  row_names_font_size <span class="op">=</span> <span class="fl">10</span>,
  show_column_names <span class="op">=</span> <span class="cn">TRUE</span>,
  cluster_columns <span class="op">=</span> <span class="cn">TRUE</span>,
  show_heatmap_legend <span class="op">=</span> <span class="cn">TRUE</span>,
  <span class="va">...</span>
<span class="op">)</span>

<span class="fu">get_fold_change_heatmap</span><span class="op">(</span>
  <span class="va">x</span>,
  <span class="va">sample_comparisons</span>,
  <span class="va">genes</span>,
  repair_genes <span class="op">=</span> <span class="cn">FALSE</span>,
  color_default <span class="op">=</span> <span class="cn">TRUE</span>,
  col <span class="op">=</span> <span class="cn">NULL</span>,
  show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,
  cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>,
  show_row_dend <span class="op">=</span> <span class="cn">TRUE</span>,
  row_names_font_size <span class="op">=</span> <span class="fl">10</span>,
  show_column_names <span class="op">=</span> <span class="cn">TRUE</span>,
  cluster_columns <span class="op">=</span> <span class="cn">TRUE</span>,
  show_heatmap_legend <span class="op">=</span> <span class="cn">TRUE</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>an abject of class "parcutils". This is an output of the function <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code>.</p></dd>
<dt>samples</dt>
<dd><p>a character vector denoting sample names to show in the heatmap.</p></dd>
<dt>genes</dt>
<dd><p>a character vector denoting genes to show in the heatmap.</p></dd>
<dt>repair_genes</dt>
<dd><p>logical, default FALSE, indicating whether to repair gene names or not. See details.</p></dd>
<dt>convert_log2</dt>
<dd><p>logical, default FALSE, indicating whether to convert gene expression values in log2 or not.</p></dd>
<dt>color_default</dt>
<dd><p>logical, default TRUE, indicating whether to use default heatmap colors.</p></dd>
<dt>col</dt>
<dd><p>an output of <code><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html" class="external-link">circlize::colorRamp2()</a></code>, default NULL.</p></dd>
<dt>convert_zscore</dt>
<dd><p>logical, default TRUE, indicating whether to convert gene expression values in to the z-score or not. see details.</p></dd>
<dt>summarise_replicates</dt>
<dd><p>logical, default TRUE, indicating whether to summarise values for each gene across replicates.</p></dd>
<dt>summarise_method</dt>
<dd><p>a character string, either mean or median.</p></dd>
<dt>show_row_names</dt>
<dd><p>logical, default FALSE, indicating whether to show row names in the heatmap or not.
Internally this argument is passed to the function <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>.</p></dd>
<dt>cluster_rows</dt>
<dd><p>logical, default TRUE, indicating whether to cluster rows in the heatmap or not.
Internally this argument is passed to the function <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>.</p></dd>
<dt>show_row_dend</dt>
<dd><p>logical, default TRUE, indicating whether to show dendrogram in the heatmap or not.
Internally this argument is passed to the function <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>.</p></dd>
<dt>row_names_font_size</dt>
<dd><p>a numeric value, default 10, indicating size of row names in the heatmap.</p></dd>
<dt>show_column_names</dt>
<dd><p>logical, default TRUE, indicating whether to show column names in the heatmap or not.
Internally this argument is passed to the function <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>.</p></dd>
<dt>cluster_columns</dt>
<dd><p>logical, default TRUE, indicating whether to cluster columns in the heatmap or not.
Internally this argument is passed to the function <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>.</p></dd>
<dt>show_heatmap_legend</dt>
<dd><p>logical, default TRUE, indicating whether to show heatmap legend or not.</p></dd>
<dt>...</dt>
<dd><p>Other parameters to be passes to <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>.</p></dd>
<dt>sample_comparisons</dt>
<dd><p>a character vector denoting sample comparisons for which heatmap of log2 fold-change to be plotted.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>an output of the function <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">ComplexHeatmap::Heatmap()</a></code>.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function <code>get_gene_expression_heatmap</code> is to create a heatmap either for normalised gene expression or z-score values
while the function <code>get_fold_change_heatmap</code> is to create a heatmap for log2 fold-changes values.</p><ul><li><p><code>repair_genes</code> :  Internally gene names are stored as a "gene_id:gene_symbol" format. For example, "ENSG00000187634:SAMD11".
When <code>repair_genes</code> is set to <code>TRUE</code> the string corresponding to gene_id followed by ":" will be removed. This is useful when gene names
to be revealed in the heatmap.</p></li>
<li><p><code>convert_zscore</code> :  When set to <code>TRUE</code> values for each gene is converted into z-score. z-score is calculated by baser r function <code><a href="https://rdrr.io/r/base/scale.html" class="external-link">base::scale()</a></code>
with all default parameters.</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

