<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="This function performs DEseq2 analysis for retained introns which are identified
by IRFinder-S program. It mainly requires two inputs - 1) output file(s) from IRFinder-S program,
and 2) sample group information to inform DEseq2 which samples to be grouped for differential expression analysis.
Other arguments are to filter retained and differently expressed introns."><title>Run DEseq2 for retained introns. — run_deseq_analysis_ir • parcutils</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run DEseq2 for retained introns. — run_deseq_analysis_ir"><meta property="og:description" content="This function performs DEseq2 analysis for retained introns which are identified
by IRFinder-S program. It mainly requires two inputs - 1) output file(s) from IRFinder-S program,
and 2) sample group information to inform DEseq2 which samples to be grouped for differential expression analysis.
Other arguments are to filter retained and differently expressed introns."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">parcutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cparsania/parcutils/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Run DEseq2 for retained introns.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cparsania/parcutils/blob/HEAD/R/ir_realted.R" class="external-link"><code>R/ir_realted.R</code></a></small>
      <div class="d-none name"><code>run_deseq_analysis_ir.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function performs DEseq2 analysis for retained introns which are identified
by IRFinder-S program. It mainly requires two inputs - 1) output file(s) from IRFinder-S program,
and 2) sample group information to inform DEseq2 which samples to be grouped for differential expression analysis.
Other arguments are to filter retained and differently expressed introns.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">run_deseq_analysis_ir</span><span class="op">(</span>
  <span class="va">x</span>,
  min_intron_cov <span class="op">=</span> <span class="fl">0.7</span>,
  min_intron_depth <span class="op">=</span> <span class="fl">10</span>,
  minimum_splice_exact <span class="op">=</span> <span class="fl">5</span>,
  minimum_replicates_with_ri <span class="op">=</span> <span class="fl">2</span>,
  <span class="va">sample_info</span>,
  group_numerator <span class="op">=</span> <span class="st">"treatment"</span>,
  group_denominator <span class="op">=</span> <span class="st">"control"</span>,
  cutoff_lfc <span class="op">=</span> <span class="fl">1</span>,
  cutoff_pval <span class="op">=</span> <span class="fl">0.05</span>,
  cutoff_padj <span class="op">=</span> <span class="fl">0.01</span>,
  regul_based_upon <span class="op">=</span> <span class="fl">1</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>x</dt>
<dd><p>an object of the class IRFinder-S. Usually this can be generated through the function <code><a href="read_irfinderS_output.html">read_irfinderS_output()</a></code>.</p></dd>
<dt>min_intron_cov</dt>
<dd><p>a value between 0 and 1, default 0.7, denoting minimum threshold for the intron coverage.
Introns with coverage less than the value will not be considered for DE analysis.</p></dd>
<dt>min_intron_depth</dt>
<dd><p>an integer, default 10, denoting threshold for minimum number of aligned reads within an intron.
Same value will be passed to argument <code>min_counts</code> in <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code> for differential analysis.</p></dd>
<dt>minimum_splice_exact</dt>
<dd><p>an integer, default 5, denoting threshold for minimum number of spliced reads within an intron.</p></dd>
<dt>minimum_replicates_with_ri</dt>
<dd><p>an integer, default 2, denoting threshold for number of replicates have an intron identified as "retained intron".
Same value will be passed to argument <code>min_replicates</code> in <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code> for differential analysis.</p></dd>
<dt>sample_info</dt>
<dd><p>internally passed to <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code>.</p></dd>
<dt>group_numerator</dt>
<dd><p>internally passed to <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code>.</p></dd>
<dt>group_denominator</dt>
<dd><p>internally passed to <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code>.</p></dd>
<dt>cutoff_lfc</dt>
<dd><p>internally passed to <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code>.</p></dd>
<dt>cutoff_pval</dt>
<dd><p>internally passed to <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code>.</p></dd>
<dt>cutoff_padj</dt>
<dd><p>internally passed to <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code>.</p></dd>
<dt>regul_based_upon</dt>
<dd><p>internally passed to <code><a href="run_deseq_analysis.html">run_deseq_analysis()</a></code>.</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="va">example_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span> ,<span class="st">"ir_data"</span> , package<span class="op">=</span><span class="st">"parcutils"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">example_files</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="va">example_dir</span>, glob <span class="op">=</span> <span class="st">"*-IR-nondir*.txt"</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">example_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span><span class="va">example_files</span> , pattern <span class="op">=</span> <span class="st">".*/"</span>, replacement <span class="op">=</span> <span class="st">""</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span class="r-in"><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"-IR-nondir.txt"</span>, replacement <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span class="r-in"><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="read_irfinderS_output.html">read_irfinderS_output</a></span><span class="op">(</span>files <span class="op">=</span> <span class="va">example_files</span>,  add_prefix_chr <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span class="r-in"><span class="va">ir_sample_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">example_files</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span class="r-in"><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="va">.</span> , groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"treatment"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">)</span></span>
<span class="r-in"><span class="fu">run_deseq_analysis_ir</span><span class="op">(</span>x <span class="op">=</span> <span class="va">x</span>, sample_info <span class="op">=</span> <span class="va">ir_sample_info</span>,</span>
<span class="r-in">min_intron_depth <span class="op">=</span> <span class="fl">2</span>, regul_based_upon <span class="op">=</span><span class="fl">1</span>, cutoff_pval <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running DESeq2 ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> converting counts to integer mode</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating size factors</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> estimating dispersions</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> gene-wise dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mean-dispersion relationship</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> final dispersion estimates</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fitting model and testing</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Done.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Summarizing DEG ...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of RI used for DE analysis are 31864. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Total number of comparisons are 1. </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Number of DE RI in each comparison:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="font-weight: bold;">treatment_VS_control</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> • <span style="color: #BB0000;">number of up RI   : 251.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> • <span style="color: #00BB00;">number of down RI : 304.</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> ──────────────────────────────</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Chirag Parsania.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

