<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="DESeq2 is a popular method to perform DE analysis for RNA-seq data. Pre and post DESeq run, however,
involves several data wrangling steps. For example, prior to run DESeq genes may need to be filtered out based on
number of reads mapped to genes across replicates. Similarly,  post DESeq run user needs to set cutoffs (log2fc, pvalue and padj)
to define up and down regulated genes. For large RNA-seq experiments involving several DE comparison such as sub-setting
and different cutoffs creates messy and less readable code. This function helps to make things bit tidy and increase
the code readability. Besides that,  output of this function can subsequently used for several other functions of this package."><title>Run differential expression (DE) analysis for several comparisosns at a time. Internally it uses
DESeq2::DESeq() to perform DE analysis. — run_deseq_analysis • parcutils</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run differential expression (DE) analysis for several comparisosns at a time. Internally it uses
DESeq2::DESeq() to perform DE analysis. — run_deseq_analysis"><meta property="og:description" content="DESeq2 is a popular method to perform DE analysis for RNA-seq data. Pre and post DESeq run, however,
involves several data wrangling steps. For example, prior to run DESeq genes may need to be filtered out based on
number of reads mapped to genes across replicates. Similarly,  post DESeq run user needs to set cutoffs (log2fc, pvalue and padj)
to define up and down regulated genes. For large RNA-seq experiments involving several DE comparison such as sub-setting
and different cutoffs creates messy and less readable code. This function helps to make things bit tidy and increase
the code readability. Besides that,  output of this function can subsequently used for several other functions of this package."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">parcutils</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cparsania/parcutils/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Run differential expression (DE) analysis for several comparisosns at a time. Internally it uses
<code><a href="https://rdrr.io/pkg/DESeq2/man/DESeq.html" class="external-link">DESeq2::DESeq()</a></code> to perform DE analysis.</h1>
      <small class="dont-index">Source: <a href="https://github.com/cparsania/parcutils/blob/HEAD/R/rnaseq_related.R" class="external-link"><code>R/rnaseq_related.R</code></a></small>
      <div class="d-none name"><code>run_deseq_analysis.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>DESeq2 is a popular method to perform DE analysis for RNA-seq data. Pre and post DESeq run, however,
involves several data wrangling steps. For example, prior to run DESeq genes may need to be filtered out based on
number of reads mapped to genes across replicates. Similarly,  post DESeq run user needs to set cutoffs (log2fc, pvalue and padj)
to define <code>up</code> and <code>down</code> regulated genes. For large RNA-seq experiments involving several DE comparison such as sub-setting
and different cutoffs creates messy and less readable code. This function helps to make things bit tidy and increase
the code readability. Besides that,  output of this function can subsequently used for several other functions of this package.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">run_deseq_analysis</span><span class="op">(</span>
  <span class="va">counts</span>,
  <span class="va">column_geneid</span>,
  <span class="va">column_samples</span>,
  <span class="va">sample_info</span>,
  <span class="va">group_numerator</span>,
  <span class="va">group_denominator</span>,
  delim <span class="op">=</span> <span class="st">"\t"</span>,
  comment_char <span class="op">=</span> <span class="st">"#"</span>,
  min_counts <span class="op">=</span> <span class="fl">10</span>,
  min_replicates <span class="op">=</span> <span class="fl">1</span>,
  cutoff_lfc <span class="op">=</span> <span class="fl">1</span>,
  cutoff_pval <span class="op">=</span> <span class="fl">0.05</span>,
  cutoff_padj <span class="op">=</span> <span class="fl">0.01</span>,
  regul_based_upon <span class="op">=</span> <span class="fl">1</span>,
  <span class="va">...</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>counts</dt>
<dd><p>a character string of the path to a count file or an object of dataframe having raw counts for each gene.
See details below to know more about format of the count file and count dataframe.</p></dd>
<dt>column_geneid</dt>
<dd><p>a character string denoting a column of geneid in <code>counts</code></p></dd>
<dt>column_samples</dt>
<dd><p>a character vector denoting names of sample columns from <code>counts</code></p></dd>
<dt>sample_info</dt>
<dd><p>a character string denoting a name of sample information file or a dataframe.
A file or a dataframe both must have at least two columns without column names. First column denotes to samples names
and second column denotes group name for each sample in first column. For e.g.</p><table class="table table"><tr><td></td><td></td></tr><tr><td>sample 1</td><td>WT</td></tr><tr><td>sample 2</td><td>WT</td></tr><tr><td>sample 3</td><td>KO</td></tr><tr><td>sample 4</td><td>KO</td></tr></table></dd>
<dt>group_numerator</dt>
<dd><p>a character vector denoting sample groups to use in numerator to calculate fold change.</p></dd>
<dt>group_denominator</dt>
<dd><p>a character vector denoting sample groups to use in denominator to calculate fold change.</p></dd>
<dt>delim</dt>
<dd><p>a character denoting deliminator for <code>count</code> file. Only valid if <code>count</code> is a file path.</p></dd>
<dt>comment_char</dt>
<dd><p>a character denoting comments line in count file. Only valid if <code>count</code> is a file path.</p></dd>
<dt>min_counts</dt>
<dd><p>a numeric value, default 10,  denoting minimum counts for a gene to be used o consider a
gene for differential expression analysis.</p></dd>
<dt>min_replicates</dt>
<dd><p>a numeric value, default 1, denoting minimum samples within a group must have <code>minimum_counts</code>.
Value provided must not be higher than number of replicates in each group.
For example for given values <code>min_replicates = 2</code> and  <code>minimum_counts = 10</code>
the genes which have minimum counts 10 in at least 2 sample groups will be used for DEG.</p></dd>
<dt>cutoff_lfc</dt>
<dd><p>a numeric value which is internally passed to <a href="categorize_diff_genes.html">categorize_diff_genes</a></p></dd>
<dt>cutoff_pval</dt>
<dd><p>a numeric value which is internally passed to <a href="categorize_diff_genes.html">categorize_diff_genes</a></p></dd>
<dt>cutoff_padj</dt>
<dd><p>a numeric value which is internally passed to <a href="categorize_diff_genes.html">categorize_diff_genes</a></p></dd>
<dt>regul_based_upon</dt>
<dd><p>either of 1, 2 or 3 which is internally passed to <a href="categorize_diff_genes.html">categorize_diff_genes</a></p></dd>
<dt>...</dt>
<dd><p>for future use</p></dd>
</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a data frame of DESeq results, DEG, and DEG summary.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>// TO DO : Explain formats of count file and count dataframe.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">count_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"toy_counts.txt"</span> , package <span class="op">=</span> <span class="st">"parcutils"</span><span class="op">)</span></span>
<span class="r-in"><span class="va">count_data</span> <span class="op">&lt;-</span> <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_delim</a></span><span class="op">(</span><span class="va">count_file</span>, delim <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">sample_info</span> <span class="op">&lt;-</span> <span class="va">count_data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">.</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span class="r-in"> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span>samples <span class="op">=</span> <span class="va">.</span> , groups <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span> ,<span class="st">"treatment1"</span> , <span class="st">"treatment2"</span><span class="op">)</span>, each <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">run_deseq_analysis</span><span class="op">(</span>counts <span class="op">=</span> <span class="va">count_data</span> ,</span>
<span class="r-in">                         sample_info <span class="op">=</span> <span class="va">sample_info</span>,</span>
<span class="r-in">                         column_geneid <span class="op">=</span> <span class="st">"gene_id"</span> ,</span>
<span class="r-in">                         group_numerator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"treatment1"</span>, <span class="st">"treatment2"</span><span class="op">)</span> ,</span>
<span class="r-in">                         group_denominator <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span><span class="op">)</span>,</span>
<span class="r-in">                         column_samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control_rep1"</span>, <span class="st">"treat1_rep1"</span>, <span class="st">"treat2_rep1"</span>, <span class="st">"control_rep2"</span>, <span class="st">"treat1_rep2"</span>, <span class="st">"treat2_rep2"</span>, <span class="st">"control_rep3"</span>, <span class="st">"treat1_rep3"</span>, <span class="st">"treat2_rep3"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="va">res</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## all comparisons</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">comp</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## DESEq result object(s)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">dsr</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## DESEq result data frame</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">dsr_tibble</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## DESEq result data frame  DEG assigned, look at the columns 'signif' and 'regul'</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">dsr_tibble_deg</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">## DEG summary</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">deg_summmary</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="op">}</span></span>
<span class="r-in"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

    </footer></div>

  

  

  </body></html>

